pallets/parachain-system/src/validate_block/mod.rs

```diff
@@ -27,7 +27,32 @@ mod tests;
 pub use polkadot_parachain;
 #[cfg(not(feature = "std"))]
 #[doc(hidden)]
-pub use sp_io;
+pub use sp_std;
+#[cfg(not(feature = "std"))]
+#[doc(hidden)]
+pub use bytes;
+#[cfg(not(feature = "std"))]
+#[doc(hidden)]
+pub use codec::decode_from_bytes;
 #[cfg(not(feature = "std"))]
 #[doc(hidden)]
 pub use sp_runtime::traits::GetRuntimeBlockType;
+
+/// Basically the same as [`ValidationParams`](polkadot_parachain::primitives::ValidationParams),
+/// but a little bit optimized for our use case here.
+///
+/// `block_data` and `head_data` are represented as [`bytes::Bytes`] to make them reuse
+/// the memory of the input parameter of the exported `validate_blocks` function.
+///
+/// The layout of this type must match exactly the layout of
+/// [`ValidationParams`](polkadot_parachain::primitives::ValidationParams) to have the same
+/// SCALE encoding.
+#[derive(codec::Decode)]
+#[cfg_attr(feature = "std", derive(codec::Encode))]
+#[doc(hidden)]
+pub struct MemoryOptimizedValidationParams {
```

Comment:
Any reason why not just directly change the `ValidationParams` to hold `Bytes` instead of `Vec<u8>`s?

(This is fine by me, I'd just like to know why because I haven't checked myself how annoying it would be to change `ValidationParams`.)

Reply:
I would have needed to change Polkadot for some implementation/optimization detail of the validation function in Cumulus. Then there would need to be some kind of comment that we don't change this etc. I mean it is possible, but then I decided to keep the changes local and observable. Also prevents that this accidentally is reverted or whatever.

Reply:
Since now we're using `decode_from_bytes` on those fields I think we'd be safe from an accidental revert, but yeah, as long as there's a test that the two don't go out of sync it's fine.

Reply:
> Since now we're using `decode_from_bytes` on those fields I think we'd be safe from an accidental revert, but yeah, as long as there's a test that the two don't go out of sync it's fine.

What I meant by accidental revert is that we "loose" the memory optimizations, nothing really breaking.